src/puya/parse.py

```diff
@@ -248,21 +256,78 @@ def parse_and_typecheck(paths: Sequence[Path], mypy_options: mypy.options.Option
             )
-            for module in EMBEDDED_MODULES.values()
+            for module in _MYPY_EMBEDDED_MODULES.values()
         ]
     )
     result = _mypy_build(mypy_build_sources, mypy_options, mypy_fscache)
     missing_module_names = {s.module_name for s in sources} - result.manager.modules.keys()
     if missing_module_names:
         # Note: this shouldn't happen, provided we've successfully disabled the mypy cache
-        raise Exception(f"mypy parse failed, missing modules: {', '.join(missing_module_names)}")
+        raise InternalError(
+            f"mypy parse failed, missing modules: {', '.join(missing_module_names)}"
+        )
+
+    # order modules by dependency, and also sanity check the contents
+    ordered_modules = []
+    for scc_module_names in mypy.build.sorted_components(result.graph):
+        for module_name in scc_module_names:
+            module = result.manager.modules.get(module_name)
+            if module is None:
+                raise InternalError(f"mypy failed to parse: {module_name}")
+            if module_name != module.fullname:
+                raise InternalError(
+                    f"mypy parsed wrong module, expected '{module_name}': {module.fullname}"
+                )
+            if not module.path:
+                raise InternalError(f"No path for module: {module_name}")
+
+            if embedded_src := _MYPY_EMBEDDED_MODULES.get(module_name):
+                module._fullname = embedded_src.puya_module_name  # noqa: SLF001
+                module_path = embedded_src.path
+            else:
+                module_path = Path(module.path)
+
+            if module_path.is_dir():
+                # this module is a module directory with no __init__.py, ie it contains
+                # nothing and is only in the graph as a reference
+                pass
+            else:
+                _check_encoding(mypy_fscache, module_path)
+                ordered_modules.append(module)
+
     return ParseResult(
         sources=sources,
         manager=result.manager,
-        graph=result.graph,
+        ordered_modules=ordered_modules,
     )
 
 
+def _check_encoding(mypy_fscache: mypy.fscache.FileSystemCache, module_path: Path) -> None:
+    module_rel_path = make_path_relative_to_cwd(module_path)
+    module_loc = SourceLocation(file=str(module_path), line=1)
+    try:
+        source = mypy_fscache.read(str(module_path))
+    except OSError:
+        logger.warning(
+            f"Couldn't read source for {module_rel_path}",
+            location=module_loc,
+        )
+        return
+    # below is based on mypy/util.py:decode_python_encoding
+    # check for BOM UTF-8 encoding
+    if source.startswith(b"\xef\xbb\xbf"):
+        return
+    # otherwise look at first two lines and check if PEP-263 coding is present
+    encoding, _ = mypy.util.find_python_encoding(source)
+    if encoding != "utf8":
+        logger.warning(
+            f"UH OH SPAGHETTI-O's,"
```

Comment:
This needs to be its own base `Error`

Reply:
It's a warning because we correctly handle the encoding in terms of reading in the code, but any unicode values (e.g. attribute names or string literals) that also get put onto the chain (such as through state keys) _might_ have different encoded bytes on-disk vs on-chain.
The spaghetti was left in at request of some folks, the message regarding compound types is getting improved/removed in another PR (#104)

Reply:
Sorry, I forgot to add the silly face emoji 😝

it was a joke… though I do think we should be able to raise an UhOhSpaghettiOError 😀

Reply:
Spaghetti-o's are srs bznss m8 😤